sender.go

```diff
@@ -122,10 +123,23 @@ func (sender *Sender) start() {
 			var header [2 * binary.MaxVarintLen64]byte
 			i := binary.PutUvarint(header[:], uint64(msg.number))
 			i += binary.PutUvarint(header[i:], uint64(flags))
-			buffer.Write(header[:i])
-			buffer.Write(body)
-			sender.conn.Write(buffer.Bytes())
-			buffer.Reset()
+			frameBuffer.Write(header[:i])
+
+			frameEncoder.enableCompression(msg.Compressed())
+			frameEncoder.write(body)
+			if msg.Compressed() {
+				// Remove the '00 00 FF FF' trailer from the deflated block
```

Comment:
Based on these comments in https://github.com/golang/go/blob/master/src/compress/flate/deflate.go#L721

```
// Flush flushes any pending data to the underlying writer.
// It is useful mainly in compressed network protocols, to ensure that
// a remote reader has enough data to reconstruct a packet.
// Flush does not return until the data has been written.
// Calling Flush when there is no pending data still causes the Writer
// to emit a sync marker of at least 4 bytes.
// If the underlying writer returns an error, Flush returns that error.
//
// In the terminology of the zlib library, Flush is equivalent to Z_SYNC_FLUSH.
func (w *Writer) Flush() error {
```

Should this check whether the bytes truncated exactly match '00 00 FF FF' trailer?  In case it wrote  '00 00 FF FF FF', in which it chop off `00 FF FF FF` and then re-add `00 00 FF FF`

> at least 4 bytes

Does that guarantee to *end* with `00 00 FF FF`?